### PR TITLE
Add variables %{lib-private} and %{libexec-private}

### PR DESCRIPTION
**Pull request by snowleopard**
This allows to find paths to source files of public and private libraries.

### CODE REVIEW

File: test/blackbox-tests/test-cases/lib/run.t [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+  >  (alias find-a)
+  >  (action (echo "%{lib:private_lib:a.ml}")))
+  > EOF
+
+  $ ./sdune build @find-a
+  File "dune", line 3, characters 18-39:
+  3 |  (action (echo "%{lib:private_lib:a.ml}")))
+                        ^^^^^^^^^^^^^^^^^^^^^
+  Error: Library "private_lib" is not public. You can use the "lib" variable
+  only with public libraries that are installed.
```
Comment (diml) [inline test/blackbox-tests/test-cases/lib/run.t]: This makes me think that it is possible to have public libraries that are not installed.

Comment (snowleopard) [reply test/blackbox-tests/test-cases/lib/run.t]: Changed to: 

```
Error: The library "private_lib" is not public. The variable "lib" expands to
the file's installation path which is not defined for private libraries.
```

Hopefully, it is clearer.

**Review by ghost — APPROVED** [review]
Looks good!

File: src/dune/expander.ml [hunk 2]
```diff
@@ -371,23 +371,41 @@ let expand_and_record acc ~map_exe ~dep_kind ~expansion_kind
...
-    match
-      Artifacts.Public_libs.file_of_lib t.lib_artifacts ~loc ~lib:lib_dep ~file
+      if lib_private then
+        let open Result.O in
+        let* lib = Lib.DB.resolve (Scope.libs t.scope) (loc, lib) in
+        let current_project_name = Scope.name t.scope
+        and referenced_project_name =
+          Lib.info lib |> Lib_info.status |> Lib_info.Status.project_name
+        in
+        if Some current_project_name = referenced_project_name then
```
Comment (aalekseyev) [inline src/dune/expander.ml]: It's probably better to avoid polymorphic equality here, since Project_name is abstract.

**Review by aalekseyev — APPROVED** [review]

File: test/blackbox-tests/test-cases/lib/run.t (hunk 1 above)
Comment (aalekseyev) [reply test/blackbox-tests/test-cases/lib/run.t]: I'm hitting this error 1.5 years later and I'm confused: why do we say the library is private even though it's actually public? At the very least the error message is confusing, but I can't even tell why it's an error in the first place.

**Pull request merged**